src/test/java/com/artipie/gem/GemITCase.java

```diff
@@ -0,0 +1,160 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.gem;
+
+import com.artipie.ArtipieServer;
+import com.artipie.RepoConfigYaml;
+import com.artipie.RepoPerms;
+import com.artipie.asto.Key;
+import com.artipie.asto.Storage;
+import com.artipie.asto.fs.FileStorage;
+import com.artipie.asto.test.TestResource;
+import com.artipie.test.RepositoryUrl;
+import com.artipie.test.TestContainer;
+import java.io.IOException;
+import java.nio.file.Path;
+import org.cactoos.text.Base64Encoded;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.core.IsEqual;
+import org.hamcrest.core.StringContains;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.EnabledOnOs;
+import org.junit.jupiter.api.condition.OS;
+import org.junit.jupiter.api.io.TempDir;
+
+/**
+ * Integration tests for Gem repository.
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ * @since 0.13
+ */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+@EnabledOnOs({OS.LINUX, OS.MAC})
+final class GemITCase {
+
+    /**
+     * Repo.
+     */
+    private static final String REPO = "my-gem";
+
+    /**
+     * Rails gem.
+     */
+    private static final String RAILS = "rails-6.0.2.2.gem";
+
+    /**
+     * Temporary directory for all tests.
+     * @checkstyle VisibilityModifierCheck (3 lines)
+     */
+    @TempDir
+    Path tmp;
+
+    /**
+     * Tested Artipie server.
+     */
+    private ArtipieServer server;
+
+    /**
+     * Container.
+     */
+    private TestContainer cntn;
+
+    /**
+     * Repository url.
+     */
+    private RepositoryUrl url;
+
+    /**
+     * Storage.
+     */
+    private Storage storage;
+
+    @Test
+    public void gemPushWorks() throws Exception {
+        final boolean anonymous = true;
+        this.init(anonymous);
+        this.push(anonymous);
+        MatcherAssert.assertThat(
+            this.storage.exists(
+                new Key.From("repos", GemITCase.REPO, "gems", GemITCase.RAILS)
+            ).toCompletableFuture().join(),
+            new IsEqual<>(true)
+        );
+    }
+
+    @Test
+    void gemInstallPushedGemWorks() throws Exception {
+        final boolean anonymous = true;
+        this.init(anonymous);
+        this.push(anonymous);
+        MatcherAssert.assertThat(
+            this.cntn.execStdout("gem", "install", GemITCase.RAILS, "--ignore-dependencies"),
+            new StringContains("Successfully installed rails-6.0.2.21")
+        );
+    }
+
+    @AfterEach
+    void tearDown() {
+        this.server.stop();
+        this.cntn.close();
+    }
+
+    private void push(final boolean anonymous) throws Exception {
+        final String key = new Base64Encoded("usr:pwd").asString();
```

Comment:
@genryxy could you please explain how `key` is used? We build it from arbitrary data, but I do not see how this data is translated into user on server. Does the test fails if auth is enabled on server, but the key is not provided on push?

Reply:
@olegmoz yes, it fails because the `gem push` expects that the gem server always requires credentials. Please, see https://stackoverflow.com/questions/61384113/how-to-push-a-gem-without-auth I'll replace `usr:pwd` with `any:any` to make it little clearer.